sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+
+    uint32_t other_field_bits = dev().GarbageMemory<uint32_t>() &
+                                ~(attr_field.bit_mask << attr_field.bit_offset);
+
+    uint32_t existing_reg_bits =
+        other_field_bits | (existing_attrs << attr_field.bit_offset);
+
+    EXPECT_READ32(attr_field.reg_offset, existing_reg_bits);
+
+    return existing_reg_bits;
+  }
+
+  /**
+   * Attribute field update kinds
+   */
+  enum AttrFieldUpdateKind {
+    kAttrFieldUpdateEnable,  /**< Enable attributes corresponding to set bits */
+    kAttrFieldUpdateDisable, /**< Disable attributes corresponding to set bits
+                              */
+    kAttrFieldUpdateWrite,   /**< Set attributes to provided bits */
+  };
+
+  /**
+   * Expects a read and write sequence required to update attributes
+   *
+   * @param attr_field Which attribute field we expect to update
+   * @param existing_attrs What existing attributes will be seen in the
+   * initial read
+   * @param update_attrs Attribute update to apply, how these are applied
+   * depends upon `update_kind`
+   * @param update_kind What kind of update to do
+   */
+  void ExpectAttrFieldUpdate(const AttrField &attr_field,
+                             dif_padctrl_attr_t existing_attrs,
+                             dif_padctrl_attr_t update_attrs,
+                             AttrFieldUpdateKind update_kind) {
+    // Check the attributes we want to update are valid
+    EXPECT_EQ(update_attrs & ~attr_field.bit_mask, 0);
+
+    ExpectLockCheck(false);
+
+    // Expect a read that will give us the existing attributes, retaining the
+    // bits the that will be returned from that read.
+    uint32_t existing_reg_bits =
+        ExpectAttrFieldRead(attr_field, existing_attrs);
+
+    uint32_t attr_field_update_bits = update_attrs << attr_field.bit_offset;
+
+    // Update the bits from the read to include the attribute update.
+    uint32_t new_reg_bits;
+
+    switch (update_kind) {
+      case kAttrFieldUpdateEnable:
+        new_reg_bits = existing_reg_bits | attr_field_update_bits;
+        break;
+      case kAttrFieldUpdateDisable:
+        new_reg_bits = existing_reg_bits & ~attr_field_update_bits;
+        break;
+      case kAttrFieldUpdateWrite:
+        new_reg_bits = (existing_reg_bits &
+                        ~(attr_field.bit_mask << attr_field.bit_offset)) |
+                       attr_field_update_bits;
+        break;
+    }
+
+    // Expect a write to perform the attribute update.
+    EXPECT_WRITE32(attr_field.reg_offset, new_reg_bits);
+  }
+
+  /**
+   * Tests various attribute enables for success
+   *
+   * @param attr_field The attribute field to test
+   * @param pad_kind What kind of pad `attr_field` is for
+   */
+  void TestEnableAttrSuccess(const AttrField &attr_field,
+                             dif_padctrl_pad_kind_t pad_kind) {
+    DeviceInit();
+
+    dif_padctrl_attr_t initial_attrs =
+        dif_padctrl_attr_t(kDifPadctrlAttrIOInvert | kDifPadctrlAttrWeakDrive);
```

Comment:
Do not use C-style casts in C++; they cause all sorts of problems, not to mention that this looks very confusingly like a function call (I was worried you were calling some kind of implicit constructor).

Use `static_cast<dif_padctrl_attr_t>(...)` instead (or define a helper, if that's too unreadable).

Reply:
Fair enough, I felt the function-style cast looked neater but point taken about the nature of C casts in C++. I have switched to `static_cast`.